src/SDKs/Azure.Base/data-plane/Azure.Base/HttpPipelineUriBuilder.cs

```diff
@@ -0,0 +1,212 @@
+                    _pathAndQuery.Append(value);
+                }
+            }
+        }
+
+        private bool HasQuery => _queryIndex != -1;
+
+        private int QueryLength => HasQuery ? _pathAndQuery.Length - _queryIndex : 0;
+
+        private int PathLength => HasQuery ? _queryIndex : _pathAndQuery.Length;
+
+        public string PathAndQuery => _pathAndQuery.ToString();
+
+        public Uri Uri
+        {
+            get
+            {
+                if (_uri == null)
+                {
+                    _uri = new Uri(ToString());
+                }
+                return _uri;
+            }
+            set
+            {
+                Scheme = value.Scheme;
+                Host = value.Host;
+                Port = value.Port;
+                Path = value.AbsolutePath;
+                Query = value.Query;
+                _uri = value;
+            }
+        }
+
+        public void AppendQuery(string name, string value)
+        {
+            ResetUri();
+            if (!HasQuery)
+            {
+                _pathAndQuery.Append(QuerySeparator);
+                _queryIndex = _pathAndQuery.Length;
+            }
+            else if (!(QueryLength == 1 && _pathAndQuery[_queryIndex] == QuerySeparator))
+            {
+                _pathAndQuery.Append('&');
+            }
+
+            _pathAndQuery.Append(name);
+            _pathAndQuery.Append('=');
+            _pathAndQuery.Append(value);
+        }
+
+        public void AppendPath(string value)
+        {
+            if (string.IsNullOrEmpty(value))
+            {
+                return;
+            }
+
+            ResetUri();
+            int startIndex = 0;
+            if (PathLength == 1 && _pathAndQuery[0] == PathSeparator && value[0] == PathSeparator)
+            {
+                startIndex = 1;
+            }
+            if (HasQuery)
+            {
+                _pathAndQuery.Insert(_queryIndex, value.Substring(startIndex, value.Length - startIndex));
+                _queryIndex += value.Length;
+            }
+            else
+            {
+                _pathAndQuery.Append(value, startIndex, value.Length - startIndex);
+            }
+        }
+
+        public override string ToString()
+        {
+            var stringBuilder = new StringBuilder();
+            stringBuilder.Append(Scheme);
+            stringBuilder.Append("://");
+            stringBuilder.Append(Host);
+            if (!HasDefaultPortForScheme)
+            {
+                stringBuilder.Append(':');
+                stringBuilder.Append(Port);
+            }
+
+            if (_pathAndQuery.Length == 0 || _pathAndQuery[0] != PathSeparator)
+            {
+                stringBuilder.Append(PathSeparator);
+            }
+
+            // TODO: Escaping can be done in-place
+            if (!HasQuery)
+            {
+                stringBuilder.Append(Uri.EscapeUriString(_pathAndQuery.ToString()));
```

Comment:
Is there an advantage of using this one, instead of `EscapeDataString`?

Reply:
Very good point. `EscapeDataString` should be used here. Changed and added a test to verify.

Reply:
I was wrong, `EscapeDataString` escapes more then we need (for example it escapes `/`)